src/Lean/Meta/Tactic/TryThis.lean

```diff
@@ -0,0 +1,585 @@
+register_option format.inputWidth : Nat := {
+  defValue := inputWidth
+  descr := "ideal input width"
+}
+
+/-- Get the input width specified in the options -/
+def getInputWidth (o : Options) : Nat := format.inputWidth.get o
+
+/-! # `Suggestion` data -/
+
+-- TODO: we could also support `Syntax` and `Format`
+/-- Text to be used as a suggested replacement in the infoview. This can be either a `TSyntax kind`
+for a single `kind : SyntaxNodeKind` or a raw `String`.
+
+Instead of using constructors directly, there are coercions available from these types to
+`SuggestionText`. -/
+inductive SuggestionText where
+  /-- `TSyntax kind` used as suggested replacement text in the infoview. Note that while `TSyntax`
+  is in general parameterized by a list of `SyntaxNodeKind`s, we only allow one here; this
+  unambiguously guides pretty-printing. -/
+  | tsyntax {kind : SyntaxNodeKind} : TSyntax kind → SuggestionText
+  /-- A raw string to be used as suggested replacement text in the infoview. -/
+  | string : String → SuggestionText
+  deriving Inhabited
+
+instance : ToMessageData SuggestionText where
+  toMessageData
+  | .tsyntax stx => stx
+  | .string s => s
+
+instance {kind : SyntaxNodeKind} : CoeHead (TSyntax kind) SuggestionText where
+  coe := .tsyntax
+
+instance : Coe String SuggestionText where
+  coe := .string
+
+namespace SuggestionText
+
+/-- Pretty-prints a `SuggestionText` as a `Format`. If the `SuggestionText` is some `TSyntax kind`,
+we use the appropriate pretty-printer; strings are coerced to `Format`s as-is. -/
+def pretty : SuggestionText → CoreM Format
+  | .tsyntax (kind := kind) stx => ppCategory kind stx
+  | .string text => return text
+
+/- Note that this is essentially `return (← s.pretty).prettyExtra w indent column`, but we
+special-case strings to avoid converting them to `Format`s and back. -/
+/-- Pretty-prints a `SuggestionText` as a `String` and wraps with respect to the pane width,
+indentation, and column, via `Format.prettyExtra`. If `w := none`, then
+`w := getInputWidth (← getOptions)` is used. Raw `String`s are returned as-is. -/
+def prettyExtra (s : SuggestionText) (w : Option Nat := none)
+    (indent column : Nat := 0) : CoreM String :=
+  match s with
+  | .tsyntax (kind := kind) stx => do
+    let w ← match w with | none => do pure <| getInputWidth (← getOptions) | some n => pure n
+    return (← ppCategory kind stx).pretty w indent column
+  | .string text => return text
+
+end SuggestionText
+
+/--
+Style hooks for `Suggestion`s. See `SuggestionStyle.error`, `.warning`, `.success`, `.value`,
+and other definitions here for style presets. This is an arbitrary `Json` object, with the following
+interesting fields:
+* `title`: the hover text in the suggestion link
+* `className`: the CSS classes applied to the link
+* `style`: A `Json` object with additional inline CSS styles such as `color` or `textDecoration`.
+-/
+def SuggestionStyle := Json deriving Inhabited, ToJson
```

Comment:
I'm actually confused again now about the need for unstructured JSON on the server side. Wouldn't it be nicer to do the style computations below in the client (which would allow for alternative clients, though that is probably not a priority)? Or do we want the style be open to extensibility?

Reply:
This is for extensibility. The previous version had a variant for all of the `def`s below it, including the rainbow colored one, and it seemed better to just let clients set their own colors if they want to customize to that extent.